freegs/reconstruction.py

```diff
@@ -0,0 +1,976 @@
+                if self.CurrentInitialisation:
+                    jtor_1d = self.initialise_plasma_current()
+                    jtor_2d = np.reshape(jtor_1d, (self.nx, self.ny),order='F')
+
+                if self.FittingWeight:
+                    F = self.get_F()
+
+            else:
+
+                if self.show:
+                    # Plot state of plasma equilibrium
+                    if pause < 0:
+                        fig = plt.figure()
+                        axis = fig.add_subplot(111)
+                    else:
+                        axs[0].clear()
+                        axs[1].clear()
+
+                    plotEquilibrium(self, axis=axs[0], show=False)
+                    if self.Jtor is not None:
+                        axs[1].imshow(np.rot90(jtor_2d), extent=(0.1,2,-1,1), cmap=plt.get_cmap('jet'))
+
+                    if pause < 0:
+                        # Wait for user to close the window
+                        plt.show()
+                    else:
+                        # Update the canvas and pause
+                        # Note, a short pause is needed to force drawing update
+                        axs[0].figure.canvas.draw()
+                        axs[1].figure.canvas.draw()
+                        plt.pause(pause)
+
+                # Use B & C to calculate Jtor matrix
+                #assumes vertical control is on
+                if self.is_limited:
+                    B = self.apply_mask(B, mask)
+                if self.VesselCurrents:
+                    jtor_1d = B @ self.c[:-(len(self.tokamak.coils) + self.Gvessel.shape[1])]
+                else:
+                    jtor_1d = B @ self.c[:-(len(self.tokamak.coils))]
+
+                jtor_2d = np.reshape(jtor_1d, (self.nx, self.ny),order='F')
+
+            # Calculate Psi values with elliptical solver
+            x_2d, mask = self.get_x(jtor_2d)
+            self.x = x_2d.flatten('F')
+
+            # Calculate B and apply mask
+            B = self.get_B()
+            if not self.is_limited:
+                B = self.apply_mask(B, mask)
+
+            # Calculating operator matrix A
+            A = self.Gplasma @ B
+            E = self.get_E(A)
+
+            # Performing least squares calculation for c
+            if self.FittingWeight:
+                self.c = scipy.linalg.lstsq(F @ E, F @ self.M)[0]
+            else:
+                self.c = scipy.linalg.lstsq(E, self.M)[0]
+
+            # Updating the Coil Currents and Vessel Currents
+            self.update_currents()
+
+            # Take Diagnostics and Perform Convergence Test
+            self.H = E @ self.c
+
+            """Not sure if this should be included or not"""
+
+            # if self.VerticalControl and self.it > 0:
+            #    for i, val in enumerate(self.H):
+            #        self.H[i] += E[i, self.pprime_order + self.ffprime_order] * self.c[self.pprime_order + self.ffprime_order] * (-1)
```

Comment:
i thought we discussed this in person
contribution from dz can be included if it reduces chi squared
should tend to small value
should only effect plasma current filaments
can discuss more in person